python/src/aac/parser/_parse_source.py

```diff
@@ -27,92 +31,62 @@ def parse(source: str, source_uri: Optional[str] = None) -> list[Definition]:
-    """Parse an Architecture-as-Code YAML file.
+def _parse_file(arch_file: str) -> list[Definition]:
+    """Parse an Architecture-as-Code YAML file and return the definitions in it.
 
     Args:
         arch_file (str): The Architecture-as-Code YAML file to be parsed.
 
     Returns:
-        The parse method returns a dict of each root type defined in the Arch-as-Code spec. If
-        validation of the provided YAML file fails, an error message is displayed and None is
-        returned.
+        The AaC definitions extracted from the specified file.
     """
-    parsed_models: dict[str, dict] = {}
+    definitions: list[Definition] = []
 
-    files = _get_files_to_process(arch_file)
+    for file in _get_files_to_process(arch_file):
+        definitions.extend(_parse_str(file, _read_file_content(file)))
 
-    for file in files:
-        contents = _read_file_content(file)
-        parsed_models = parsed_models | _parse_str(file, contents)
+    return definitions
 
-    return parsed_models
 
-
-def _parse_str(source: str, model_content: str) -> dict[str, dict]:
+def _parse_str(source: str, model_content: str) -> list[Definition]:
     """Parse a string containing one or more YAML model definitions.
 
     Args:
         source:  The file the content came from (to help with better logging)
         model_content:  The YAML to parse
 
     Returns:
-        A dictionary of the parsed model(s). The key is the type name from the model and the
-        value is the parsed model root.
+        The AaC definitions that were built from the model contents.
     """
-    parsed_models: dict[str, dict] = {}
 
-    roots = _parse_yaml(source, model_content)
+    def mark_to_source_location(start: yaml.error.Mark, end: yaml.error.Mark) -> SourceLocation:
+        return SourceLocation(start.line, start.column, start.index, end.column - start.column)
+
+    def get_lexemes_for_definition(contents):
+        yaml_source = path.abspath(source) if path.lexists(source) else DEFAULT_SOURCE_URI
+        lexemes = []
+        tokens = [token for token in yaml.scan(contents, Loader=yaml.SafeLoader) if hasattr(token, "value")]
+        for token in tokens:
+            location = mark_to_source_location(token.start_mark, token.end_mark)
+            lexemes.append(Lexeme(location, yaml_source, token.value))
+        return lexemes
 
-    for root in roots:
-        if "import" in root:
-            del root["import"]
+    definitions: list[Definition] = []
+    for doc in model_content.split(YAML_DOCUMENT_SEPARATOR):
+        for root in _parse_yaml(source, doc):
+            if "import" in root:
+                del root["import"]
 
-        root_type, *_ = root.keys()
-        root_name = root.get(root_type).get("name")
-        parsed_models = parsed_models | {root_name: {source: root}}
+            root_type, *_ = root.keys()
+            root_name = root.get(root_type).get("name")
+            contents = _add_yaml_document_separator(doc)
```

Comment:
We can't assume that every definition has a preceding yaml doc separator. We should use regex to split the definitions without destructively removing yaml separators, however I wasn't able to quickly identify a regex that was really capable of accomplishing that.

Reply:
I don't really see this assumption here so I'm not really sure I see the problem you're pointing out. If there is no document separator in the string/file contents we would add it (which is what was desired based on the AC). If there is no document separator and there should be one (i.e. there are two AaC definitions in the same YAML document) the user would eventually get an error to that effect since we're checking for that all over the place.

Can you give an example of a YAML document that would show what the issue is? In that way, at the very least I can create a test that confirms this works how we want it to.

Reply:
With the assumption you're errantly adding document separators to the initial definition of files like gen_protobuf.yaml. https://github.com/jondavid-black/AaC/blob/c8bc393ab7432d376f0c660a5de7a280329df10f/python/src/aac/plugins/gen_protobuf/gen_protobuf.yaml#L1-L43

As for the AC, the intention is to only add preceding doc separators that are present in the file, not to introduce content that's not present in the source file.